tree/dataframe/src/RLoopManager.cxx

```diff
@@ -813,10 +814,15 @@ void RLoopManager::Jit()
       }
    }
 
-   const std::string code = []() {
-      R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
-      return std::move(GetCodeToJit());
-   }();
+   R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
```

Comment:
Consider the following code:
```
   R__READ_LOCKGUARD(ROOT::gCoreMutex);
   if (GetCodeToJit().empty()) {
      R__LOG_INFO(RDFLogChannel()) << "Nothing to jit and execute.";
      return;
   }

   R__WRITE_LOCKGUARD(ROOT::gCoreMutex);
   // Check again if another thread has already cleared the global string
   // with the code to JIT. Without this check, we could end up calling
   // InterpreterCalc with an empty string, which would raise an exception.
   if (GetCodeToJit().empty()) {
      R__LOG_INFO(RDFLogChannel()) << "Nothing to jit and execute.";
      return;
   }
   const std::string code = std::move(GetCodeToJit());
```

Reply:
Just for my own understanding, your suggestion is almost equal to the current code, except that the first `R__READ_LOCKGUARD` is not enclosed in a scope but is at the "global" function scope. Why is it better to create a read lock first, then create a write lock in the same scope? The situation that should occur is that multiple threads may reach the read lock part, check for the emptiness of the string, then move to the write lock part where only one thread at a time can continue. But I don't understand how writing the read lock at function scope is different than writing it within an inner scope

Reply:
The difference is that with the inner scope the operation are:

1. Take Read Lock
2. Give away Read lock
3. Take Write Lock

While without it, it is

1. Take Read Lock
2. Switch To Write Lock

So it reduces the lock churn, especially if there is no contention.

Reply:
Ah, cool then, thanks!